truffle/src/main/java/org/jruby/truffle/stdlib/readline/ReadlineNodes.java

```diff
@@ -0,0 +1,310 @@
+        }
+
+    }
+
+    @CoreMethod(names = "insert_text", constructor = true, required = 1)
+    @NodeChildren({
+            @NodeChild(type = RubyNode.class, value = "self"),
+            @NodeChild(type = RubyNode.class, value = "text")
+    })
+    public abstract static class InsertTextNode extends CoreMethodNode {
+
+        @CreateCast("text") public RubyNode coerceTextToString(RubyNode text) {
+            return NameToJavaStringNodeGen.create(null, null, text);
+        }
+
+        @TruffleBoundary
+        @Specialization
+        public DynamicObject insertText(DynamicObject readline, String text) {
+            getContext().getConsoleHolder().getReadline().getCursorBuffer().write(text);
+
+            return readline;
+        }
+
+    }
+
+    @CoreMethod(names = "delete_text", constructor = true)
+    public abstract static class DeleteTextNode extends CoreMethodArrayArgumentsNode {
+
+        @TruffleBoundary
+        @Specialization
+        public DynamicObject deleteText(DynamicObject readline) {
+            getContext().getConsoleHolder().getReadline().getCursorBuffer().clear();
+
+            return readline;
+        }
+
+    }
+
+    @CoreMethod(names = "line_buffer", onSingleton = true)
+    public abstract static class LineBufferNode extends CoreMethodArrayArgumentsNode {
+
+        @Child private TaintNode taintNode;
+
+        public LineBufferNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+            taintNode = TaintNodeGen.create(context, sourceSection, null);
+        }
+
+        @TruffleBoundary
+        @Specialization
+        public Object lineBuffer() {
+            final CursorBuffer cb = getContext().getConsoleHolder().getReadline().getCursorBuffer();
+
+            final DynamicObject ret = createString(StringOperations.createRope(cb.toString(), getDefaultInternalEncoding()));
+            return taintNode.executeTaint(ret);
+        }
+
+    }
+
+    @CoreMethod(names = "refresh_line", onSingleton = true)
+    public abstract static class RefreshLineNode extends CoreMethodArrayArgumentsNode {
+
+        @TruffleBoundary
+        @Specialization
+        public DynamicObject refreshLine() {
+            try {
+                getContext().getConsoleHolder().getReadline().redrawLine();
+            } catch (IOException e) {
+                throw new RaiseException(coreExceptions().ioError(e.getMessage(), this));
+            }
+
+            return nil();
+        }
+
+    }
+
+    // Taken from org.jruby.ext.readline.Readline.ProcCompleter.
+    // Complete using a Proc object
+    public static class ProcCompleter implements Completer {
+
+        private final DynamicObject procCompleter;
+
+        //\t\n\"\\'`@$><=;|&{(
+        static private String[] delimiters = {" ", "\t", "\n", "\"", "\\", "'", "`", "@", "$", ">", "<", "=", ";", "|", "&", "{", "("};
```

Comment:
I'm a bit confused why this is an array of String, since it seems only character are allowed, and conversion seem to happen all the time to a flat String. Is it maybe to handle surrogate pairs? Anyway, it's OK as is for now, just confusing.